core/vm/logger.go

```diff
@@ -43,5 +43,52 @@ type EVMLogger interface {
 	CaptureFault(pc uint64, op OpCode, gas, cost uint64, scope *ScopeContext, depth int, err error)
 	CaptureKeccakPreimage(hash common.Hash, data []byte)
 	// Misc
-	OnGasConsumed(gas, amount uint64)
+	OnGasChange(old, new uint64, reason GasChangeReason)
 }
+
+// GasChangeReason is used to indicate the reason for a gas change, useful
+// for tracing and reporting.
+type GasChangeReason byte
+
+const (
+	GasChangeUnspecified GasChangeReason = iota
+
+	// GasInitialBalance is the initial balance for the call which will be equal to the gasLimit of the call
+	GasInitialBalance
+	// GasRefunded is the amount of gas that will be refunded to the caller for data returned to the chain
+	// PR Review: Is that the right description? Called in core/state_transition.go#StateTransition.refundGas
+	GasRefunded
```

Comment:
`GasRefunded` is the sum of all refunds which happened during the tx execution (e.g. storage slot being cleared).

I see the total that will be returned to the caller is tracked via `GasBuyBack`. Is this distinction important? I would think if we want to track the refund values tracking each of them is more valuable than the sum at the end? i.e. emit an event at the site when storage is cleared etc.

Reply:
> I see the total that will be returned to the caller is tracked via GasBuyBack. Is this distinction important?

Good question actually. I do feel its important but it's a personal opinion I'm sharing here as we didn't track those before.

My thinking here is that gas refund has been used speculatively for "gas value accrual" in the past so emitting this event make sense to me, it will be easy to know how much refund has been given back to you.

If it was "blended" with `GasBuyBack`, I'm unsure it would be possible to extract the refund value. I might be wrong and maybe. it could be computed/inferred based on surrounding gas changes, I do still think an explicit version is better.

I do get that it could be blended in indeed.

> I would think if we want to track the refund values tracking each of them is more valuable than the sum at the end

I don't think so, here is why. Correct me if I'm wrong, when the storage is cleared and the refund accumulate, the gas counter is not incremented meaning the gas value the execution sees don't contain the refund.

It's would be falsely IMO to emit a `GasChange` at those points because the execution don't account for them yet. I think the logger should faithfully trace the execution as it happens.

Let me know what do you think.

Reply:
> the gas counter is not incremented meaning the gas value the execution sees don't contain the refund.

Ah yes that's a good point.

> I think the logger should faithfully trace the execution as it happens.

Agreed.

>  it will be easy to know how much refund has been given back to you.

Ok then we keep a separate reason for refunds 👍